### PR TITLE
[aballesta] Create cache file

### PR DESCRIPTION
**Pull request by aballesta**

### CODE REVIEW

File: lib/puppet/functions/hiera_aws_sm.rb [hunk 1]
```diff
@@ -48,6 +61,19 @@ def lookup_key(key, options, context)
...
 
+    # Handle cache
+    if options.key?('cache_ttl')
+      cache_ttl = options['cache_ttl']
+    else
+      cache_ttl = 24
+    end
+
+    if options.key?('cache_file')
+      cache_file = options['cache_file']
```
Comment (cmartinez85) [inline lib/puppet/functions/hiera_aws_sm.rb]:
```suggestion
      cache_file = options['cache_file']
      log.info("Using cache #{cache_file}")
```

**Review by aitor-bodas — APPROVED** [review]

File: lib/puppet/functions/hiera_aws_sm.rb [hunk 2]
```diff
@@ -166,6 +195,64 @@ def secret_key_name(key, options)
...
+        end
+      end
+      return false
+    else 
+      return false
+    end
+  end
+
+  def file_cache_exist(options)
+    two_hours_ago = DateTime.now - (options['cache_ttl']/24.0)
```
Comment (cmartinez85) [inline lib/puppet/functions/hiera_aws_sm.rb]:
```suggestion
    cache_expiration_time = DateTime.now - (options['cache_ttl']/24.0)
```

File: lib/puppet/functions/hiera_aws_sm.rb [hunk 3]
```diff
@@ -166,6 +195,64 @@ def secret_key_name(key, options)
...
+
+    if(File.exist?("#{options['cache_file']}"))
+      if (two_hours_ago.strftime( "%Y-%m-%d %H:%M:%S" ) > File.ctime("#{options['cache_file']}").strftime( "%Y-%m-%d %H:%M:%S" ))
```
Comment (cmartinez85) [inline lib/puppet/functions/hiera_aws_sm.rb]:
```suggestion
      if (cache_expiration_time.strftime( "%Y-%m-%d %H:%M:%S" ) > File.ctime("#{options['cache_file']}").strftime( "%Y-%m-%d %H:%M:%S" ))
```

File: lib/puppet/functions/hiera_aws_sm.rb [hunk 4]
```diff
@@ -166,6 +195,64 @@ def secret_key_name(key, options)
...
+         generate_cache(options['cache_file'], options['prefixes'])
+      end
+    else
+         generate_cache(options['cache_file'], options['prefixes'])
+    end
+  end
+
+  def generate_cache(path, prefixes)
+    client = Aws::SecretsManager::Client.new(
+      region: 'eu-west-1',
```
Comment (cmartinez85) [inline lib/puppet/functions/hiera_aws_sm.rb]: We are passing the region from hiera.yaml so you don't need to hardcode this region there

File: lib/puppet/functions/hiera_aws_sm.rb [hunk 5]
```diff
@@ -166,6 +195,64 @@ def secret_key_name(key, options)
...
+    if(File.exist?("#{options['cache_file']}"))
+      if (cache_expiration_time.strftime( "%Y-%m-%d %H:%M:%S" ) > File.ctime("#{options['cache_file']}").strftime( "%Y-%m-%d %H:%M:%S" ))
+         generate_cache(options['cache_file'], options['prefixes'])
+      end
+    else
+         generate_cache(options['cache_file'], options['prefixes'])
+    end
+  end
+
+  def generate_cache(path, prefixes, options)
```
Comment (cmartinez85) [inline lib/puppet/functions/hiera_aws_sm.rb]: Are prefixes included in options?

Comment (cmartinez85) [inline lib/puppet/functions/hiera_aws_sm.rb]:
```suggestion
  def generate_cache(path, prefixes, region)
```

File: lib/puppet/functions/hiera_aws_sm.rb [hunk 6]
```diff
@@ -166,6 +195,64 @@ def secret_key_name(key, options)
...
+  end
+
+  def generate_cache(path, prefixes, region)
+    client = Aws::SecretsManager::Client.new(
+      region: region,
+    )
+    resp = client.list_secrets({
+      max_results: 100,
+    })
+        file = File.open(path, 'w')
```
Comment (bjuncosa) [inline lib/puppet/functions/hiera_aws_sm.rb]: Fix indentation, please

**Review by bjuncosa — APPROVED** [review]

**Review by cmartinez85 — APPROVED** [review]

**Review by cmartinez85 — APPROVED** [review]

**Review by bjuncosa — APPROVED** [review]

**Pull request merged**